server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -983,16 +983,33 @@ public void writeWriteable(Writeable writeable) throws IOException {
     }
 
     public void writeException(Throwable throwable) throws IOException {
-        writeException(throwable, throwable, 0);
+        writeException(throwable, throwable);
     }
 
-    private void writeException(Throwable rootException, Throwable throwable, int nestedLevel) throws IOException {
+    private static final ThreadLocal<Integer> CURRENT_NESTED_LEVEL = ThreadLocal.withInitial(() -> 0);
```

Comment:
I am not sure why ThreadLocal is needed. Can you add a comment why it had to be used? 
do we expect this method to be run concurrently?

Reply:
The `ThreadLocal` is needed to maintain nested level context through multiple method calls that can't be updated to add an `int` tracking parameter. It's `ThreadLocal` because multiple threads could be calling this code at the same time, and they all need to maintain independent nesting counts